### ISSUE TITLE
Documentation format

### ISSUE THREAD

**Issue by kanta** [ponoor/python-step-series#5]
Hi again @JulianOrteil ,
I am trying to port my markdown docs to reStructuredText, so far "[Functional descriptions](https://ponoor.com/en/docs/step-series/functional-description/)" part.

In markdown, we can use an inline code and a link at the same time, like this;
[`/setVoltageMode`](https://ponoor.com/en/docs/step-series/osc-command-reference/voltage-and-current-mode-settings/#setvoltagemode_intmotorid)

But it seems not possible with reStructredText. How do you think, should we avoid to use this sort of expressions in the library's docs?

Thanks!
kanta

**Comment by JulianOrteil** [comment]
@kanta https://sublime-and-sphinx-guide.readthedocs.io/en/latest/references.html#links-to-external-web-pages

I recommend using the separated link and target syntax. Targets (`.. _CNN: http://cnn.com/` from the example) will be used for any recurrence of the link syntax: ``` `CNN`_ ```. Put targets as the last thing in your documentation.

To see an example, here's the index page of the documentation: https://raw.githubusercontent.com/ponoor/python-step-series/main/docs/index.rst

Note how there are multiple occurrences of ``` `Sphinx`_ ``` but only one occurrence of ``` .. _Sphinx: https://www.sphinx-doc.org/```. Also note how links and targets aren't limited to just one word.

**Comment by kanta** [comment]
Thanks! Yes, I am using the separated link style. Here are some working files (not finished).
https://github.com/ponoor/step-series-support/tree/main/docs/functional-descriptions/rst

In the original Markdown, I added a link to an inline code like `/exampleMsg`. Inline code + link is working in Markdown but not in reStructuredText. Headings + link as well. So I was wondering if there is a standard way to do that.
I only found the way using an extension of sphinx. If so, these kind of excerptions are not good for portability so probably I would remove them.

Let me ask one more thing, it seems there are many way to express a headings. I am converting as following so far.
```
##    : =
###   : -
####  : ^
##### : '
```
Example:
https://raw.githubusercontent.com/ponoor/step-series-support/main/docs/functional-descriptions/rst/homing-and-position-management_en.rst

If you have any other recommendation, please let me know.
Thank you!

**Comment by JulianOrteil** [comment]
@kanta I opened [sss#12](https://github.com/ponoor/step-series-support/pull/12) so you can compare your current syntax errors against what it is supposed to be. Feel free to accept or reject at your leisure.

> Headings + link as well. So I was wondering if there is a standard way to do that.

There should be no difference in syntax for external links in headers:
```
====================
`Sensor and switch`_
====================

.. _Sensor and switch: https://www.google.com/
```
This should work just fine.

> I only found the way using an extension of sphinx.

I have never had to use an extension for Sphinx beyond ``rtd_theme``. Most extensions are typically edge-case or useful in very limited circumstances--and usually, there's a way around them. 😋 



> Let me ask one more thing, it seems there are many ways to express a heading. I am converting as following so far.

In reST, there is no "standardized" heading list like in markdown (which is denoted by the amount of #'s you have) until you start writing your documentation. Whatever heading format list you start with should remain consistent throughout all of your documentation. Because I've already written large portions of the documentation, this is the list you should use:
```
#    : #
##   : *
###  : =
#### : -
##### : ^
```
I've only reached as far as the headings in this list. If you need to go further, use whichever symbol you see fit next.

**Comment by JulianOrteil** [comment]
As always, reach out and I'll do my best to help 👍 

Jules

**Comment by kanta** [comment]
@JulianOrteil Thank you for help! I will check the diff and correct a syntax. Also thanks for the heading list, I will work with it!

**Comment by kanta** [comment]
@JulianOrteil I corrected the syntax of "Functional description" part.
https://github.com/ponoor/step-series-support/tree/main/docs/functional-descriptions/rst

I will start to work for "Technical information" part but I will update the content first, since some descriptions are not up to date.

**Comment by JulianOrteil** [comment]
@kanta It looks good! When you're ready, submit these to the library and I'll review them for English grammar mistakes. I may combine your docs into one file to try and group them under one heading.

Also, please look at my comments on our discussion board about translations when you have the opportunity.

**Comment by kanta** [comment]
@JulianOrteil Thanks for the check! I updated some docs and ported to rst files with the same manner.
https://github.com/ponoor/python-step-series/tree/documentation/docs/portedDocs

Please review them when you have time!

**Comment by JulianOrteil** [comment]
@kanta I am reviewing your documentation now. I've thus far completely restructured the docs to make them easier to manage and to better future proof them for if you would like to develop another motor control board. These changes are currently live.

Also, while reading through your docs, I noticed you state `the STEP400 has an upper rating limit of 5A due to the power rating limitation of the current sensing resistor`. At my day job, I'm a software developer for a variety of different electronic devices (which now looks to soon include actual robotics thanks to your motor controllers). We heavily rely on the INA260 for measuring power. While it has a maximum voltage rating of 36V, its current rating is 15A. We use them all the time with Arduino thanks to Adafruit: https://www.adafruit.com/product/4226

I thought I'd bring them to your attention if you weren't aware of them already as a possible workaround to this limitation.

Jules

**Comment by JulianOrteil** [comment]
@kanta Functionality Descriptions have been reviewed. They are currently live: https://python-step-series.readthedocs.io/en/latest/step-series/about.html#functionality-descriptions

**Comment by kanta** [comment]
@JulianOrteil Thank you very much for your review, reconstruction and update, happy to see this organized result! I found some my typos in external PDFs in the repo, also noticed a link to the schematics for both models are missing in the doc. So I have updated about them and pushed [them](https://github.com/ponoor/python-step-series/commit/af07aaf684c44302a9ad3329563f10e25ef7ba20).

And thank you for your idea of INA260. A current detection with a resistor is hardcoded in the driver chip PowerSTEP01 (= the MCU is outside of the feedback loop and unfortunately i2c won't work) and actually we can extend the current rating up to 10A with a smaller resistance resistor. My original thought was that 10A for an each phase x4 motors could be too much for most of components (a wire width of the pcb, reverse polarity protection MOSFET, power terminal, etc..) so 5A could be a reasonable physical limitation, otherwise the PCB became quite massive.
At the same time, in the real projects, I always wanted to see what happen with a higher current. It's always preferable to have a enough room for the configurable range so probably this could be a good ToDo for the next version.
Thank you for your idea!

**Comment by kanta** [comment]
By the way I'm curious what you are doing in the daytime. If your project with step-series become released to public, please let me know!!

**Comment by JulianOrteil** [comment]
Surely! What DuraTech intends (the company I work for) won't be public, so I can't share the specifics about it. All I can say is the machine the board might control will be instrumental in making displays for medical equipment, washer and dryers, and other devices.

My personal project is to make a 1:2 or 1:4 scale version of GLaDOS from Portal 2. Kind of like this: https://www.youtube.com/watch?v=96riTrt-BbU

Also, my week this week is looking to be pretty much packed. I dont know when I'll be able to review your next docs. Probably next week sometime.

**Comment by kanta** [comment]
Thanks for telling me your work, that sounds really professional stuffs.
Your personal projects also sounds like really serious!

By the way, I found a problem on my firmware. Some bool responses from the controller are sent as integer (1/0), but some are sent as bool (T/F). IF you set `/enableBusyReport` you will get a response `/busy` width integer but if you send `/getBusy`, the response will be `/getBusy` with True/False. There are no differences in my main environment Max so I haven't been aware of it.
I would unify all those response arguments into integers. I hope this change won't impact the Python library.

**Comment by JulianOrteil** [comment]
@kanta No worries. The only impact it has on the library is the type annotations I have. I can make a change and test them so we should be a-ok.

**Comment by kanta** [comment]
Hi again, 

I implemented changes above. Now all responses with logical values should be expressed as integer (1/0). The responses of these commands have been changed:
- /setDestIp
- /getConfigName
- /getHomeSw
- /getLimitSW
- /getBusy
- /getDir
- /getHiZ
- /getUlvo
- /getHomeSwMode
- /getLimitSwMode
- /getHomingDirection
- /getProhibitMotionOnHomeSw
- /getProhibitMotionOnLimitSw

And I added description to the [doc](https://ponoor.com/en/docs/step-series/osc-command-reference/) about an OSC-type of arguments for both incoming and outgoing messages.

Also, by the past update, an executable timing of some motion commands were changed from when the doc was written. So I added one another condition "In motion start condition".
The condition of those commands are changed:
- /run
- /move
- /goTo
- /goToDir
- /goHome
- /goMark

Sorry for these changes after you reviewed/edited, but it would be nice if you can reflect those to the new doc.

I had struggled with another issue in these days and changed some for the [firmware](https://github.com/ponoor/step-series-universal-firmware) but they were internal things in the board so OSC messages should not affected.
Thanks!!

**Comment by JulianOrteil** [comment]
@kanta Sounds good. I'll get these changes implemented in the upcoming days.

Thank you for your detailed descriptions.

**Comment by JulianOrteil** [comment]
@kanta I updated the responses module of the library to only return `int`s instead of `bool`s. Please double check to make sure I didn't miss any.

**Comment by kanta** [comment]
@JulianOrteil I have checked the update and all are correct. Thank you!
I will come back to work for the translation soon.

**Comment by JulianOrteil** [comment]
@kanta Great to hear. I saw you worked on them a little bit earlier. I hope the process is super smooth throughout! :)

**Comment by kanta** [comment]
@JulianOrteil Hello Jules, 
Sorry for my slow progress but I have a question for a link translation. Links are described as \`link\`_ and it's definition is places in the end of the each documents. It will work for translation if it's a name of device, product or command, like "PowerSTEP01", "openFrameworks", or "`/setBemfParam`", since they are written with alphabet too in Japanese. But some other words won't work like "datasheet", "pre-made-connection cables", or etc... Can I use inline link for them, or do you have any idea to treat them?
Thanks!

**Comment by JulianOrteil** [comment]
@kanta
No need for apologies, these docs are yours, so the timetable for getting this done is up to you. If anything, I should be the one apologizing for slacking these last few weeks haha.

As for links, I'm actually not quite sure. Because this is my first time doing multilingual support, I haven't encountered this before. A quick Google search hasn't revealed anything of value either, so here's what I would try: add the links for the translated Japanese text to the English file.

In the .rst file:
```shell
A sentence with the link `English translation`_.

.. _English translation: https://www.example.com/
.. _Japanese translation: https://www.example.com/    <--- Would be actual Japanese, not English
```

In the .po file:
```shell
msgid ""
"A sentence with the link `English translation`_."
msgstr ""
"A sentence with the link `Japanese translation`_."   <--- Would be actual Japanese, not English
```

If that doesn't work, then try the inline method.

**Comment by JulianOrteil** [comment]
@kanta I just ported over some more technical information. I ran ``tox -e docs`` to update your .po files and some of your translated text was automatically reorganized; can you review it and check for any errors?

Also, as a reminder, please make sure to run ``tox -e docs`` before committing any changes. Ignore (don't commit) the following 3 files if they generate when you run the command: 
![image](https://user-images.githubusercontent.com/23083317/165214175-67b07a01-04ae-4ed5-a3ce-346743363bc8.png)

**Comment by JulianOrteil** [comment]
@kanta I just finished porting over the technical-related information from your website. I will be redoing some of the documentation I've already done to bring it up to speed with what we are doing now (as our methods have slightly changed since we've started). Afterwards, it will be onto ``responses.py``.

Please let me know if I missed any information or if you find any errors in my ports.

**Comment by kanta** [comment]
@JulianOrteil Ahh such a nicely organized porting, thank you very much!
I will read the new section more precisely soon and let you know if I found something.
Thanks for pointing out `tox` stuffs which I haven't recognized correctly. I will continue the porting and will commit it with that command. 
It becomes a holiday season here and my update will be slow down (since more frequent physical interruption to my PC by my son...) but will come back within 2 weeks.

**Comment by JulianOrteil** [comment]
@kanta Hope you are doing well! It is my goal to finish off the English ports this week for your documentation so you don't have to worry about waiting for me to finish something off.

I'm almost finished with the responses documentation and I notice that ``/getLimitSwMode`` is not finished. For ``swMode``, is it the same as ``/getHomeSwMode``?

![image](https://user-images.githubusercontent.com/23083317/169715274-2bb15411-bfee-41b5-b536-2a4a22f78379.png)

**Comment by kanta** [comment]
@JulianOrteil Ah you are correct, `swMode` is the same as `/getHomeSwMode`. 
Thank you for your hard work! I was bit stacked these weeks but I will catch up your progress (step by step).

**Comment by JulianOrteil** [comment]
@kanta Excellent, then ``responses`` should be finished. Ill get the electromagnetic brake started tonight and then finish sometime soon, hopefully this week.

**Comment by JulianOrteil** [comment]
@kanta Would you be willing to provide a picture for connecting the EM brake to the motor? Your English description is a bit hard to follow and I believe users would benefit from seeing how you're expecting the wiring to be done.

If not, I'll do my best to clarify your descriptions.

Otherwise, all documentation has been completed on my end. Please let me know if I missed something and I'll put it in!

**Comment by kanta** [comment]
@JulianOrteil Sure, here some pics.
https://ponoor.com/cms/wp-content/uploads/2021/03/brake-power.jpg
https://ponoor.com/cms/wp-content/uploads/2021/03/step400-brake-connection.jpg
https://ponoor.com/cms/wp-content/uploads/2021/03/brake-riboncable.jpg
https://ponoor.com/cms/wp-content/uploads/2021/03/pkp243d.jpg

And I have just realized the documentation for the EM board is independent from STEP400/800, and I haven't port them to `rst` files yet.
https://ponoor.com/en/docs/emb-step400/

Probably better to combine to your new docs?

BTW I got the inquiry about how to use STEP400 with Python in Jetson nano last week. I believe this library will be very helpful for many people. Thank you again for your big contribution!

**Comment by JulianOrteil** [comment]
@kanta Thank you for the pictures, I'll make sure to put them in. As for the brake documentation, I have already finished them. They should be live and ready.

Also great to hear! Always happy to help wherever I can :)

**Comment by kanta** [comment]
@JulianOrteil Hi Julian, let me have a question.
I found quite similar contents in `python-step-series.po` and `module-reference.po`.
For example, the line 223 and subsequent lines of [python-step-series.po](https://github.com/ponoor/python-step-series/blob/4de8884668a4adf643f21ba887cd7513869219f1/docs/locales/ja/LC_MESSAGES/python-step-series.po#L223) are mostly same as line 51+ of  [module-reference.po](https://github.com/ponoor/python-step-series/blob/4de8884668a4adf643f21ba887cd7513869219f1/docs/locales/ja/LC_MESSAGES/module-reference.po#L51). Is one of them a byproduct of the automated process? If so, which one should I add the translations?
Thank you for your help!

**Comment by JulianOrteil** [comment]
@kanta This is a byproduct of the automation process, good guess!

I'm not quite sure which one is considered to be the "master" file, so perhaps doing a test will yield the answer. Try adding a small translation to one of the files, commit it, and see if it changes on ReadTheDocs?

**Comment by kanta** [comment]
@JulianOrteil Thanks! I found `python-step-series.po` is the master and the other was not affected to the result. 

During the porting process, I found some my original descriptions were not up to date or just wrong. If I edit the English .rst files (I would ask you to review my English, though), how can I continue the workflow? Just `tox -e docs` should be fine for `.po` files?

**Comment by JulianOrteil** [comment]
@kanta That is correct. ``tox -e docs`` will update the po files for you to include the new text that needs translating.

Let me know when you're finished updating the descriptions in English and I'll review them.

**Comment by JulianOrteil** [comment]
@kanta Just checking in to make sure everything is going well for you and if you need anything from me. Do let me know if something comes up and you need my assistance on something :)

**Comment by kanta** [comment]
@JulianOrteil Thanks! Still very slow progress but I hope to make some progress when I can work.
I updated [some explanations](https://github.com/ponoor/python-step-series/commit/8b82561680fc8500c7088807cd21f6f820d3183b) in `commands.py`. Could you check these text? Thanks!!

**Comment by JulianOrteil** [comment]
@kanta I have reviewed your English documentation changes and updated them.

I also patched ``SetDestIP`` to now wait for a response before returning. You should no longer need ``time.sleep`` when on the ``documentation`` branch.

**Comment by kanta** [comment]
@JulianOrteil Thank you for your review! I will continue to port the translation texts.
Your new patch seems very helpful for many users/environments, thanks a lot!

**Comment by JulianOrteil** [comment]
I'll release the patch once you finish your translations of the documentation. Then people will just get the fix when they update or install the library 🚀

**Comment by kanta** [comment]
@JulianOrteil Hi again Julien! I am currently working on completing the translation of the document, and I updated the explanations about the homing direction again. Could you please review them?
https://github.com/ponoor/python-step-series/commit/64550bc2e1a8e28f5242ffd537fc39e2fe4f8fcb
My intension is clearing the idea of the homing direction. It is a simply user predefined direction for the homing and no algorithm behind it. Even if the motor overrun the HOME or LIMIT switches/sensors, the homing direction won't be updated automatically. So I tried to use the expression with a physical position of the switch/sensor.

One more thing. It seems markup symbols in reStructuredText need space between 2-bytes-charactors, otherwise markups aren't recognized. Many commands in `Device Commands` have a section of 'alias of --' but in the translated text, 'の別名です。' is not preceded by a space so the text is not rendered properly.
![image](https://user-images.githubusercontent.com/1097902/177253227-1b24d8e5-ca4b-42c0-8882-4edefc330f26.png)

I have tried to find the definition of this translation among `.po` files but I couldn't find any. Do you know is there any predefined translations in the gettext system? If you are not sure, I will dig into it more.

Thanks a lot!

**Comment by JulianOrteil** [comment]
@kanta Apologies, I must have read through your post here and got completely caught up in other projects. Sorry about that!

I have slight grammar tweaks to your explanations I'd like to make, but I'd really like to clarify the ``HomeSwActivating`` and ``LimitSwActivating`` error notifications. Just to be clear, these are sent when movement is _requested_ in the direction the home/limit sensor is activated, right? For example, if the home sensor is activated, and the user requests the motor to ``/run`` in the direction of home, the ``HomeSwActivating`` error is returned. OR is it when either sensor activates?

As for the weird, breaking formatting, those appear to be automatically generated by sphinx. I'm not aware of any configuration we can do to fix that, so I'll see if I can find something. To be honest, it even appears to be a bug to me, so I can file a ticket with the correct repo and respond with the link to it so you can track it as well if I can't find something on my own.

Jules

**Comment by JulianOrteil** [comment]
After looking around, I could not find anything. I have made a ticket in sphinx's repo: https://github.com/sphinx-doc/sphinx/issues/10747

I am not convinced this issue is related to autodoc, but it's my best wild guess at the moment.

**Comment by kanta** [comment]
@JulianOrteil No worries, glad to know you are fine and busy! 
Yes, as you said, `HomeSwActivating` and `LimitSwActivating` errors are only sent when the movements are requested. 

Thanks for working for the sphinx formatting as well. So far any reStructuredText formats will be destroyed without a space before 2 byte characters. It'd be nice if this problem is solved!

**Comment by JulianOrteil** [comment]
I've updated the documentation with slight tweaks here and there and clarified the error messages.

**Comment by JulianOrteil** [comment]
@kanta Are there any further major changes you'd like to make to the docs? If not, I'll perform a release.

**Comment by kanta** [comment]
@JulianOrteil Great! I haven't added translations for some updated text so could you wait a bit for them? It shouldn't be a long duration. Thanks!

**Comment by JulianOrteil** [comment]
No worries. Didn't mean to rush you, I just didn't know if there was anything left needing translating. Let me know when you are ready!

**Comment by kanta** [comment]
@JulianOrteil I found if there is a sentence `#, fuzzy` in the beginning of English text in `.po` files, then the translated text won't be rendered to the Japanese docs. For example, [here ](https://github.com/ponoor/python-step-series/blob/documentation/docs/locales/ja/LC_MESSAGES/step-series.po#L1172)is one of `#, fuzzy` in `step-series.po`, and [here ](https://python-step-series.readthedocs.io/ja/latest/step-series/about.html#voltage-vs-current-mode)is the result.
Can I remove these `#, fuzzy` by hand, or if you have any idea for the sphinx setting?
Thank you for your help as usual!

**Comment by JulianOrteil** [comment]
@kanta I don't think there'd be an issue if you removed them, so go ahead and give it a shot. I'm not even sure what it does, but if it's causing issues, then I'd remove them.

**Comment by kanta** [comment]
@JulianOrteil Thanks, I removed them and it worked well!
I think I have done for the translation. 'alias of --' parts are still not right but I'd like to make it an issue for the future.

**Comment by JulianOrteil** [comment]
Okay, if everything looks good to you, then I'll make a release ready tomorrow or sometime later this week. Just give me to go ahead when you've finished any final prep you want to do.

**Comment by kanta** [comment]
I found some more missing translations and just updated them. These changes have not yet been reflected in the readthedoc, but I believe we are now ready for release!

**Comment by JulianOrteil** [comment]
@kanta I have pushed a release candidate to testpypi. If you have the opportunity, please install the updated version using the command ``pip install --index-url https://test.pypi.org/simple/ --extra-index-url https://pypi.org/simple python-step-series==1.0.0rc2`` and give it a test!

I will as well, but if you could also try this release in your environment, please do!

I'll release the prod. release by the end of the week or whenever you give the green light from your tests if you run them.

**Comment by kanta** [comment]
@JulianOrteil Thanks for the instruction. I created a new virtualenv , and installed the package as you instructed. Then it works! I made some motions and got some BUSY state and position responses. My environment is Windows11/Anaconda/Python3.9.12.
Everything seems to be working well on my end!

**Comment by JulianOrteil** [comment]
Perfect. As I am currently developing with the library for an in-house application at my workplace, I'll verify it works with Linux tomorrow and then do a release sometime soon. Thank you for the quick verification!

**Comment by JulianOrteil** [comment]
By the way, is there any documentation we are missing or do you think it's ready for you to "flip the switch" and redirect customers/clients to the new version from your website? I'd think maintaining the endpoint on your website would be best for compatibility if only it were to redirect to the new docs on RTD.

**Comment by JulianOrteil** [comment]
@kanta The tests were successful on my end. I will perform a release later today.

**Comment by kanta** [comment]
@JulianOrteil Congratulations for v1.0.0!! I'm sure this library will help users a lot.

For the new docs, there are no OSC references in it so this could be an another future issue.
So other explanation pages can be redirected to the new one.

**Comment by JulianOrteil** [comment]
@kanta Thank you, and to you as well for putting in so much work getting us this far!
